File: C/Chapter4/quickSort.c
#include <stdio.h>

void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int s[], int low, int high)
{
    int p = high;                // Pivot element index
    int firstHigh = low;        // Divider position for pivot element
    for (int i = low; i < high; i++)
    {
        if (s[i] < s[p])
        {
            swap(&s[i], &s[firstHigh]);
            firstHigh++;
        }
    }
    swap(&s[p], &s[firstHigh]);
    return firstHigh;
}


void quickSort(int array[], int low, int high)
{
    int p;
    if (low < high)
    {
        p = partition(array, low, high);
        quickSort(array, low, p - 1);
        quickSort(array, p + 1, high);
    }
}

int main()
{
    int s[10] = {11, 32, 54, 67, 2, 34, 56, 55, 90, 6};
    printf("Before sorting:\n");
    for (int i = 0; i < 10; i++) printf("%d\n", s[i]);
    quickSort(s, 0, 9);
    printf("After sorting:\n");
    for (int i = 0; i < 10; i++) printf("%d\n", s[i]);
    return 0;
}

File: Python/Chapter4/quickSort.py
from typing import List


def quick_sort(array: List[int], low: int, high: int):
    if low < high:
        partition_index: int = partition(array, low, high)
        quick_sort(array, low, partition_index - 1)
        quick_sort(array, partition_index + 1, high)


def partition(array: List[int], low: int, high: int):  # O(n)
    p = high
    first_high = low - 1
    for i in range(low, high):
        if array[i] < array[p]:
            first_high += 1
            array[i], array[first_high] = array[first_high], array[i]
    array[p], array[first_high + 1] = array[first_high + 1], array[p]
    return first_high + 1


if __name__ == '__main__':
    nums = [1, 5, 6, 3, 8, 0, 2, 2]
    quick_sort(nums, 0, len(nums) - 1)
    print(nums)
